### DISCUSSION TITLE
Merging docs and switching to github pages

### DISCUSSION THREAD

**Discussion by uuf6429** [Behat/Behat#1490] [Ideas]
The title says it all :)

### Why?

- The `Behat/docs` has been forgotten for a while, which has caused documentation to be outdated.
- It will be easier to manage/maintain.
- It will be more visible to contributors.
- PRs can be blocked until documentation has updated (just like blocking for tests to be added)
- A more obvious and natural digital trail between feature and documentation.

### Contras

- An arguably (slightly) more bloated main repository.
- Not able to have documentation-only maintainer access
- ???

### Next steps...

- With git it's possible to merge a repo into an another one while also preserving the original history. I did that in the past - there might be a bit of experimenting involved.
  - Not entirely sure, but might need admin (push --force) access to push such a drastic change.
  - Keeping the history for documentation might not be very beneficial (and the old repo would still be available anyway)
- With [`.gitattributes`](https://github.com/Behat/Behat/blob/master/.gitattributes), we would be able to exclude the docs dir from (composer) packages, so it won't affect performance.
- Existing issues/PRs (and branches) would not be ported, so we need to get rid of them first (and probably archive the repo before migrating it).

### Why?

- Less things to maintain (external service)
  - (and giving access to it to maintainers)
- A more consolidate CI/CD flow
  - Triggering documentation deployments
  - Seeing deployment failures + history etc (quite a few Issues in `Behat/docs` ask if the documentation has been released or not; or if/why it has failed)
- Easily connect deployments e.g. redeploying documentation when a new release is made.

### Contras

- GitHub has some limitations/thresholds (number of (concurrent) builds, build time etc).
- GitHub Pages is for purely static content.
  - It's possible to host a browser-based PHP VM though (e.g. [example](https://uuf6429.github.io/rune-examples/shop/))
  - AFAIK we only have static content so far anyway
- Not natively able to show a preview for PRs

### Next steps...

- we'd need the right GitHub action setup
- we'd need to enable GitHub Pages in the project setup
- when ready, someone would need to change the relevant DNS records

**Comment by ciaranmcnulty** [comment]
Makes sense to me to merge the repositories

Whether it's GH Pages or the current readthedocs.org I don't have a strong opinion; I suspect lack of knowledge about Hugo etc. could be a deciding factor?

**Reply by uuf6429** [reply]
> I suspect lack of knowledge about Hugo etc

Ah, I knew I forgot something 😄 We wouldn't need to change the system - the gh action would simply generate and upload static content. This [example repo](https://github.com/annegentle/create-demo) uses RST/sphinx and gh pages (if I'm not mistaken, we're using sphinx with readthedocs).

I would be in favour of also switching to markdown, but I don't have a strong opinion on that.

**Reply by ciaranmcnulty** [reply]
Whatever the minimal changeset to get the existing stuff working (readthedocs uses sphinx I think)

**Comment by acoulton** [comment]
Re GH pages - this was briefly discussed in https://github.com/Behat/docs/issues/129 this week. The biggest thing RTD offers that GH Pages doesn't is the ability to automatically generate previews for all open PRs, as well as for multiple release branches. Being able to see the built version of a PR would make it much easier to review.

This is not natively supported by GH Pages, which can only serve a single branch / version of a site at any one time. It could probably be done with an amount of scripting and github actions to compile all the versions into e.g. separate subdirectories within one overall Pages site. However this could be quite complex to get right (and might mean quite a slow pages build). There may be Marketplace Actions / other projects already existing that could help.

If someone wants to explore that, great. If not I suggest we stick with RTD (assuming we can get it working again).

**Reply by acoulton** [reply]
On the pros / cons list:

* I think almost all the pros you have for GH Pages would also apply to RTD if the integration was working (they will deploy automatically, post back commit statuses, etc). 
* You're right RTD is an external service, but AFAIK in theory maintainers should not need to access it often. It's only because RTD switched from the old github service authentication flow in 2019 to become (I assume) a github app and nobody has reconfigured that. Once the projects are re-linked I think it would "Just Work" (until/unless Github change their permissions model for third-party services again).
* I don't think there's any issue that GH Pages can't run PHP - nor can RTD - the content is all static once compiled.

**Reply by uuf6429** [reply]
> I think almost all the pros you have for GH Pages would also apply to RTD if the integration was working (they will deploy automatically, post back commit statuses, etc).

There's a big difference... with GH you actually see the code (config) AND that is being built & deployed. RTD can at best post some status.

> in theory maintainers should not need to access it often

That doesn't really matter. Maintainers also do not need access to git very often with the old release cycle. :). I believe that maintainers should at least be able to see if docs are released (or why they're not), just as being able to see/update the regular code release flow.

> I don't think there's any issue that GH Pages can't run PHP

I was trying to avoid having that question coming up.

**Reply by uuf6429** [reply]
> If someone wants to explore that, great. If not I suggest we stick with RTD (assuming we can get it working again).

If there is an agreement on the objectives, I'd gladly take up the challenge to explore it further.

**Reply by acoulton** [reply]
I think the discussion about RTD vs Pages is based on some misunderstandings about how it works and what it can do, which is understandable given it doesn't currently work and has not been for a long time.

With **both** github pages **and** RTD:

* They are essentially transparent to the people writing / editing / maintaining docs, and are "fire and forget" after the initial setup - which as I understand it for RTD is we just need to grant RTD permissions on the repo again and let them set up their webhook
* They do/can provide a build result back to github through a commit status
* They allow contributors & maintainers to see the build logs for every build - on Pages this is in Github Actions, on RTD it lives with the docs site e.g. https://app.readthedocs.org/projects/behat/builds/?utm_source=behat&utm_content=flyout and is accessible either from the version switcher on the page or from the link they include on the github commit status.
* They store their entire build config in the github repo - ours is [currently here](https://github.com/Behat/docs/blob/v3.0/conf.py), it looks like the project is too old / using default config to have a .readthedocs.yaml but if we want RTD to do something different (e.g. build a different framework) then the process for configuring that is identical to configuring a custom Pages build.

ReadTheDocs has the added advantage that out-of-the-box it:

* serves rendered sites for different versions of the docs at the same time, with an interface on every page of the docs to allow the end-user to see which version they are looking at and switch between them.
* serves rendered staging sites for all open pull requests, which are available as soon as the PR is submitted and linked from the PR - see for example the commit status and link from this PR on Jupyter https://github.com/jupyter/jupyter/pull/747

Github Pages / Actions *may* be more immediately familiar to some people, but not necessarily - it depends on exactly what actions / site rendering engine / workflows we're using and whether they have seen that setup before.

As I said, if someone wants to spend time investigating / working out how to set up an equivalent docs site on GH Pages then I'm not totally wedded to RTD. But it seems to me it could be quite a lot of work that could be avoided, given I can't see any feature GH Pages has that RTD doesn't have (if RTD was working).

**Comment by acoulton** [comment]
Re merging the docs and code repositories - I am not in favour:

> * The Behat/docs has been forgotten for a while, which has caused documentation to be outdated.

In my experience, docs in projects get forgotten and outdated because people aren't prompted / pushed to update them (e.g. by contributor guidelines and PR templates when they submit a PR / by maintainers when it is reviewed). Where they live is a secondary issue - I have seen plenty of projects with dead documentation sitting in a `/docs` folder in the main repo, and plenty with actively maintained docs in a separate repo.

> It will be easier to manage/maintain.

Not necessarily. For a start it would be impossible to have volunteers who were "docs maintainers" without also giving them push access to Behat/Behat - which might not be desired. That can be a very good role for getting people to engage with and support a project when they're starting out. It would create noise in the commit history (and releases, and actions workflow histories) for Behat/Behat. It would also create new potential for merge conflicts, which might be easier to solve if the code and the documentation were in separate branches in separate repos.

> It will be more visible to contributors.

This can be solved by e.g. highlighting the docs repo from the Pull Request template.

> PRs can be blocked until documentation has updated (just like blocking for tests to be added)
> A more obvious and natural digital trail between feature and documentation.

This can be solved by having a process where we ask people to submit a docs PR and link it to their code PR (where appropriate). Some projects enforce this with github actions, for example, to apply a commit status.

I suggest for the moment we would be better focusing our time on getting the docs more of a priority in the processes we already have, and getting them up to date, than moving repos, reconfiguring build tooling, etc.

**Reply by uuf6429** [reply]
> In my experience, docs in projects get forgotten and outdated because people aren't prompted / pushed to update them

Having a separate repo will only make that harder, there's no questioning that.
It is irrelevant if some projects managed to have external documentation or even no documentation at all - they might afford that when they have a large contributor base and no viable competition.

What is relevant here is asking contributors to also clone this repo to make a 1 line documentation change, making a PR, getting it reviewed and ready for release.... on top of their intended change in the main repo. The average dev will not care about that and avoid contributing in the first place.

I know because that happened to me numerous times. I can't be bothered to spend 10x more time on updating documentation - because of some weird build system, templating system, code style and a tonne of other things.

> Not necessarily. For a start it would be impossible to have volunteers who were "docs maintainers" without also giving them push access to Behat/Behat.

Documentation PRs need to be reviewed just as well. And such maintainers need to be accountable anyway. Injecting malware on the docs site is just as bad, so I'm sure we won't be handing out maintainer access to docs without a care, regardless where the repo lies.

> It would create noise in the commit history (and releases, and actions workflow histories) for Behat/Behat.

I'd argue that that "noise" is instrumental in keeping the docs up to date and relevant. Having a "documentation silo" means that even the most up-to-date documentation would still risk technical inaccuracies - and we should avoid that as much as outdated documentation.

> It would also create new potential for merge conflicts

How is that even a problem? Conflicts would happen either way. How is it that solving conflicts in 2 repos is easier than 1?

> I suggest for the moment we would be better focusing our time on getting the docs more of a priority in the processes we already have, and getting them up to date, than moving repos, reconfiguring build tooling, etc.

I never said we have to do this now. It would be more important to get some consensus.
If we agree, someone else might be able to take care of this.

**Reply by acoulton** [reply]
> What is relevant here is asking contributors to also clone this repo to make a 1 line documentation change, making a PR, getting it reviewed and ready for release.... on top of their intended change in the main repo. The average dev will not care about that and avoid contributing in the first place.

This is a reasonable point, **but** if it really is a one-line change, it can be done in the github web editor - no need to clone anything. If it's a bigger change, the effort to clone one repo is quite small in proportion to the time spent writing.

> I know because that happened to me numerous times. I can't be bothered to spend 10x more time on updating documentation - because of some weird build system, templating system, code style and a tonne of other things.

All of these points are irrelevant to what repo the docs sit in (and indeed what they're hosted on).

> Documentation PRs need to be reviewed just as well. And such maintainers need to be accountable anyway. Injecting malware on the docs site is just as bad, so I'm sure we won't be handing out maintainer access to docs without a care, regardless where the repo lies.

Sure, there's a baseline "I trust you not to inject malware" but that doesn't mean"I trust you to make and review code changes". It also doesn't follow that everyone who can change code is the right person to review docs. There might also be people willing to volunteer for docs tasks who wouldn't want to / feel confident dealing with the codebase.

> I'd argue that that "noise" is instrumental in keeping the docs up to date and relevant.

It depends on the updates. Changes to functionality, sure. But fixing broken links, fixing typos, changing the docs templating language are all things end-users don't really need to see or care about the history of.

> merge conflicts

I guess I was thinking docs conflicts are often more common than code conflicts, due to the nature / structure of files, and since github won't build a PR with a merge conflict, it would be a pain for us to have to solve conflicts before we could see if a contribution to the code had passed CI. It's a fairly small part of my concerns though.

**Reply by acoulton** [reply]
The other thing that occurred is if we inline the docs, it would make it harder to have strict branch protection rules - e.g. that some / all of the tests must be green before merging - unless we run the full CI for every contribution, but that has downsides too.

I can think of quite a few workflows / checks we might want on code changes that we wouldn't want on docs changes, and vice versa. The list of commit statuses is already long when reviewing pulls, do we really want more?

**Comment by carlos-granados** [comment]
I don't have a strong opinion on Github pages vs ReadTheDocs, my only comment is that people are probably going to be more familiar with GH pages than with RDT, so this may help if people want to contribute to the docs infrastructure.

@uuf6429 mentioned in passing the possibility of moving to Markdown. For me this would be a great plus, it would make it exponentially easier for people to contribute to the docs. But I understand this is a big change

And regarding merging both repositories, I would be in favour of this. I believe this would greatly reduce the friction for people to contribute to the documentation as part of working on a new feature

**Reply by uuf6429** [reply]
This reminds me, @acoulton the point about PR previews could be solved with switching to markdown:
https://docs.github.com/en/pages/setting-up-a-github-pages-site-with-jekyll/setting-a-markdown-processor-for-your-github-pages-site-using-jekyll

**Reply by acoulton** [reply]
@uuf6429 I don't understand that point? If you mean github can render individual markdown files in the web UI - [github can also render .rst files](https://github.com/Behat/docs/blob/v3.0/quick_start.rst). 

What [github can't do](https://github.com/orgs/community/discussions/21582) is show you the whole built version of the docs - how pages fit / link together, how they look in the template etc - for a pull request, unless we build some custom pages builder that can re-build the docs for every branch (including iterating PRs from forks) and aggregate that into one overall "site", presumably grouped by directories, every time it runs. And then feed back on the PR with a link to the directory for the branch you're on.

One GH repo = One GH pages site, with no support for partial deployments. Unless I've missed something.

**Reply by uuf6429** [reply]
> If you mean github can render individual markdown files in the web UI - [github can also render .rst files](https://github.com/Behat/docs/blob/v3.0/quick_start.rst).

I wasn't aware of that, but then I wouldn't bother about having a full site rendering preview - unless we're doing a tonne of custom stuff the GH render should be good enough (I'm assuming links would also work fine). But it's a fair counterpoint to consider.

**Comment by carlos-granados** [comment]
Seems that ReadTheDocs supports having both RST and MD content in the same project simultaneously. so perhaps a transition to Markdown could be orchestrated to take place slowly
https://docs.readthedocs.io/en/stable/guides/migrate-rest-myst.html

**Reply by uuf6429** [reply]
Cool, good to know!

**Comment by acoulton** [comment]
I'm finding the threading here a little hard to follow. We have I think 3 separate things being discussed at once:

* Having the docs hosted on Github pages instead of RTD.
* Switching to a different page-building engine instead of sphinx and/or switching from RST to markdown (both of which can be done on RTD, and both of which can be done on Github pages).
* Merging the docs into this repo (which again, could happen or not regardless of the other choices).

I've tried to put my thoughts into each of the threads.

I do still think merging the docs is a bad idea, but I'm happy to be outvoted on that.

**Reply by carlos-granados** [reply]
Yes, agree with you that we are talking about three different things here and that this makes this discussion complicated. What if we create three different issues on the repo, one to explore the possibility of merging it with the main repo, one to explore the possibility of using a different engine/markdown and another one to explore using Github pages. Then someone can take ownership/lead on each of them and explore the possibilities

But in any case I think that all this should be future work, I believe we have more urgent things to look into:
- We really need to get the integration with ReadTheDocs working again
- We need to do a review of the current content as there is a lot of information that is missing. For example, currently there is no list of command line parameters and their usage

**Reply by acoulton** [reply]
Sounds like a good plan, and completely agree on priorities.

I am on the case with RTD, I only asked about access on Friday afternoon UK time so was going to wait till end of today then bump https://github.com/Behat/docs/issues/129 if we've not had a reply.

As you say, in the short term any time anyone has to look at / think about docs would be most useful on content - even just reading through & opening issues for the things that should be added or improved. Maybe also looking at the history of 3.x and seeing if there are features that have not yet been documented.

**Reply by uuf6429** [reply]
I considered having multiple discussions, but then I figured all three points are rather interrelated, so it would make sense to have some discussion, consensus and then separate issues for any decisions.

Regarding priorities, from my perspective, we could live some more time without RTD if it means we will have a (different) solution that works in the end.
Otherwise I'm sure that after we're done spending time restoring RTD and reviewing/updating the docs the argument of "why bother redoing all of it differently" would come up.

Obviously I don't want to force this on anyone, I get that restoring RTD is probably the quickest way, given the right access. The point is that if we consider doing it differently, I could jump in to do the transition, with minimal effort/dependence on the maintainers.

But it's just my opinion on that.

**Reply by stof** [reply]
The blocker for restoring RTD is an intervention from @everzet to grant us access (and maybe an action from him or @ciaranmcnulty in case adding the webhook requires an admin to do it as other maintainers have only write access on the repo)
The blocker for migrating the doc elsewhere than RTD is being able to change the DNS, which also requires an intervention from @everzet

**Comment by AlexSkrypnyk** [comment]
Re custom docs implementation.

I've throughly reviewed https://just-the-docs.com/ and https://docusaurus.io/ and ended up [using](https://getscaffold.dev/) Docusaurus for PHP and other projects because it:
1. Supports MD and MDX (for those cases when the advanced docs are needed).
2. Widely used by the community
3. Can be used to host an entry page (like a homepage)
4. Has a lot of plugins

Example of config can be [found here](https://github.com/AlexSkrypnyk/scaffold/tree/main/docs).

GHA to build and deploy docs to GH Pages can be found [here](https://github.com/AlexSkrypnyk/scaffold/blob/main/.github/workflows/release-docs.yml).
